### DISCUSSION TITLE
JS API

### DISCUSSION THREAD

**Discussion by pca006132** [elalish/manifold#327] [General]
I'm proposing a different API style for js, integrating a 2D subsystem (#127) and provide a simple vector API (#238). The idea is to make transformations polymorphic, so they can be applied to points, lines, polygons and 3D meshes. For example:

```javascript
transforms = compose([
  translate([10, 10, 0]),
  rotate([20, 0, 0]),
])

transform([1, 2, 3]) // point transformation
transform(circle(10)) // polygon transformation
transform(cube) // mesh transformation
```

The problem is that I am not sure about what to do when the dimensions disagree: What should we do when the user tries to apply a 3D transformation to a 2D polygon? Should we drop the z axis value? Or should we keep the transformation and apply it in 3D?

And I'm thinking about what primitives should we have for polygons, maybe we can have some more primitives such as interpolating and approximating b-spline or something?

**Comment by elalish** [comment]
I think we should retain the distinction between dimensions, more like GLM with `mat4x3` for 3D points/meshes and `mat3x2` for 2D points/polygons. We could come up with our own transform composition API, but it might be simpler to use an existing one, #324.

**Comment by geoffder** [comment]
For composing transformations I think it is nice to have affine matrix versions of the same transformation primitives that are available for the shape and vector types, such that it is possible to build up a single matrix that can apply composed transformations in one go. Plus it's something that you can support without forcing everything to go through a transformation DSL.

To use something that looks like the API I use in OCaml (BOSL2 in OpenSCAD and Gg in OCaml do similar, and I'm sure there are other examples), we could have a module Affine3 with an interface that exposes these:
```ocaml
val translate : v3 -> mat4x3
val rotate : ?about:v3 -> v3 -> mat4x3
val axis_rotate : ?about:v3 -> v3 -> float -> mat4x3
val scale : v3 -> mat4x3
val compose : mat4x3 -> mat4x3 -> mat4x3
val ( %> ) : mat4x3 -> mat4x3 -> mat4x3
val ( % ) : mat4x3 -> mat4x3 -> mat4x3n
```
Compose simply being a flipped multiplication:
```ocaml
let compose a b = mul b a
let ( %> ) = compose
let ( % ) = mul
```
Then each of the transformable types have some function affine that takes the appropriate affine matrix type (2d or 3d) and applies it.
```ocaml
let a = 
  let open Affine3 in
  scale (v3 5. 2. 0.) 
  %> rotate ~about:(v3 0. 0. 5.) (v3 5. 0. 0.) 
  %> translate (v3 0. 6. 4.)
let v = Vec3.affine a (v3 1. 1. 1.)
let m = Manifold.(affine a (sphere 10.)) 
```

**Reply by pca006132** [reply]
Yes, this would be the cleanest approach. I'm thinking if we add some non-linear transformations in the future, e.g. offsetting or something, it will require something different. Although to be fair it seems that we will not have offsetting in the near future.
